### DISCUSSION TITLE
How does TPM2 MITM protection fit into Heads?

### DISCUSSION THREAD

**Discussion by JonathonHall-Purism** [linuxboot/heads#1655] [General]
With TPM2, it's possible to ensure that there is not an attacker intercepting the connection to the TPM.  (We check the primary handle against a known value when setting up the encrypted session, this is similar to SSH known_hosts.)  TPM1.2 does not have this at all, this is TPM2 only.

The current method stores a digest of the primary handle in /boot, and signs it.  But this does not seem effective against an attacker that can open the device, which is necessary to MITM the TPM.  (A device-opening attacker could gain access to the TOTP/HOTP secret, which means they could alter the firmware.  They could remove the MITM check or just intercept whatever they want from the firmware, no need to MITM the TPM.)

An alternative I can think of is to extend the TOTP/HOTP secret with the primary handle digest, so a change in the primary handle (an MITM attack) would cause TOTP/HOTP verification to fail.  (E.g. the TOTP/HOTP secret would become `sha256(unsealed_secret + prim_handle_digest)`.)  This does not require storing any trust anchor for the TPM on the device itself, essentially it's stored on the TOTP/HOTP authenticator.  But I'm also not sure whether this gains any security guarantee against a device-opening attacker.

Either way, I tend to think the primary handle digest on /boot is ineffective, so we should remove it to eliminate the headaches it causes currently.

Originated from discussion in https://github.com/linuxboot/heads/pull/1630#issuecomment-2075120429

**Comment by tlaurion** [comment]
Note that code refers to comment, not here

https://github.com/linuxboot/heads/blob/3fef9e06c1bb9de81a170e747282ab5cea879360/initrd/bin/tpmr#L360-L365


Error on console was suppressed successfully (first attempt was unsuccessful) under https://github.com/tlaurion/heads/commit/25d88f208da740cd906baca343c043e722bc6271

@JonathonHall-Purism

**Comment by tlaurion** [comment]
> Note that code refers to comment, not here
> 
> https://github.com/linuxboot/heads/blob/3fef9e06c1bb9de81a170e747282ab5cea879360/initrd/bin/tpmr#L360-L365
> 
> Error on console was suppressed successfully (first attempt was unsuccessful) under [tlaurion@25d88f2](https://github.com/tlaurion/heads/commit/25d88f208da740cd906baca343c043e722bc6271)
> 
> @JonathonHall-Purism

Concerned user raised discussion for firmware older then 3 weeks ago, didn't see the error on screen being new after upgrade of tpm2 toolstack, today at
https://matrix.to/#/!pAlHOfxQNPXOgFGTmo:matrix.org/$YGvznhSP5qAh0gK2Nbzw9KkunSUboiEMA3EsheWozUs?via=matrix.org&via=nitro.chat&via=tchncs.de

**Comment by tlaurion** [comment]
@JonathonHall-Purism to be prioritized alongside of of hotp counter eradication, see https://github.com/linuxboot/heads/issues/1873#issuecomment-2577986022 in goal of supporting unsafe usb boot

**Comment by tlaurion** [comment]
> Either way, I tend to think the primary handle digest on /boot is ineffective, so we should remove it to eliminate the headaches it causes currently.

Pr to remove primary handle https://github.com/linuxboot/heads/pull/2153
